src/mbgl/renderer/tile_pyramid.cpp

```diff
@@ -138,14 +138,12 @@ void TilePyramid::update(const std::vector<Immutable<style::Layer::Impl>>& layer
         std::unique_ptr<Tile> tile = cache.get(tileID);
         if (!tile) {
             tile = createTile(tileID);
-            if (tile) {
-                tile->setObserver(observer);
-                tile->setLayers(layers);
+            if (!tile) {
+                return nullptr;
             }
+            tile->setObserver(observer);
         }
-        if (!tile) {
-            return nullptr;
-        }
+        tile->setLayers(layers);
```

Comment:
I don't think this is the right fix -- forcing a tile to re-layout whenever it is retrieved from the cache negates most of the benefit of caching.

Reply:
Agreed it is not ideal. In this case the other option left is to discard, but not save in cache stale tiles when needing relayout. This also works and now I realize this is the approach with less performance impact. I'll update the PR accordingly.